# Data Cleaning Notebook

This notebook cleans the raw laptop review dataset based on the issues found during data understanding. 
It removes duplicate rows and fixes the data types of no_ratings and no_reviews. 
The cleaned dataset is saved separately so the raw data remains untouched.

### Step 1: Load the raw dataset

We load the original raw CSV again in this new notebook, so this notebook can run independently 
from start to finish without depending on the previous one.

In [3]:
import pandas as pd

# load the raw dataset fresh, so this notebook runs independently
df = pd.read_csv("../data/raw/laptops_dataset_final_600.csv")
print("Shape before cleaning:", df.shape)
df.head(3)

Shape before cleaning: (24113, 7)


,product_name,overall_rating,no_ratings,no_reviews,rating,title,review
0,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/...,4.7,"15,210",900,5,Perfect product!,"Loved it, it's my first MacBook that I earned ..."
1,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/...,4.7,"15,210",900,5,Fabulous!,Battery lasted longer than my first relationsh...
2,Apple MacBook AIR Apple M2 - (8 GB/256 GB SSD/...,4.7,"15,210",900,5,Fabulous!,Such a great deal.. very happy with the perfor...


### Step 2: Remove duplicate rows

During data understanding, we found 7122 duplicate rows, about 30 percent of the dataset. 
Investigation showed these were long, uniquely worded reviews repeated under the same product, 
most likely caused by the scraper capturing the same review more than once. 
Since these are not distinct opinions from different customers, it is safe to remove them.

In [4]:
print("Duplicate rows before removal:", df.duplicated().sum())

# drop exact duplicate rows, keep only the first occurrence
df = df.drop_duplicates()

print("Shape after removing duplicates:", df.shape)
print("Duplicate rows after removal:", df.duplicated().sum())

Duplicate rows before removal: 7122
Shape after removing duplicates: (16991, 7)
Duplicate rows after removal: 0


### Step 3: Fix data types for no_ratings and no_reviews

Both columns are stored as text because some values contain commas as thousands separators. 
We remove the commas and convert both columns to whole numbers.

In [5]:
# remove comma separators, then convert to integer type
df["no_ratings"] = df["no_ratings"].str.replace(",", "").astype(int)
df["no_reviews"] = df["no_reviews"].str.replace(",", "").astype(int)

df[["no_ratings", "no_reviews"]].dtypes

no_ratings    int64
no_reviews    int64
dtype: object

### Step 4: Verify the cleaned dataset

Before saving, we check the dataset one more time to confirm duplicates are gone, 
data types are correct, and there are no missing values introduced during cleaning.

In [6]:
print("Final shape:", df.shape)
print()
print("Duplicate rows:", df.duplicated().sum())
print()
print("Missing values:")
print(df.isnull().sum())
print()
print("Data types:")
print(df.dtypes)

Final shape: (16991, 7)

Duplicate rows: 0

Missing values:
product_name      0
overall_rating    0
no_ratings        0
no_reviews        0
rating            0
title             0
review            0
dtype: int64

Data types:
product_name          str
overall_rating    float64
no_ratings          int64
no_reviews          int64
rating              int64
title                 str
review                str
dtype: object


### Step 5: Save the cleaned dataset

The cleaned dataset is saved to the processed data folder, separate from the raw data. 
This keeps the original raw file untouched and allows the cleaning steps to be re-run if needed.

In [7]:
# save cleaned data, index=False avoids writing an extra unnamed column
df.to_csv("../data/processed/processed_data.csv", index=False)
print("Saved cleaned dataset with shape:", df.shape)

Saved cleaned dataset with shape: (16991, 7)


## Cleaning Summary

The following cleaning steps were applied to the raw dataset.

1. Removed 7122 exact duplicate rows, found to be caused by repeated scraping of the same reviews. 
Dataset reduced from 24113 rows to 16991 rows.

2. Converted no_ratings and no_reviews from text to whole numbers, after removing comma separators 
that were preventing pandas from reading them as numeric.

3. Verified the cleaned dataset has zero duplicates, zero missing values, and correct data types 
across all columns.

The cleaned dataset is saved as data/processed/processed_data.csv and is ready for the next stage, 
aspect extraction.